core/src/main/java/io/grpc/internal/Http2ClientStreamTransportState.java

```diff
@@ -0,0 +1,236 @@
+    if (transportError != null) {
+      // Already received a transport error so just augment it.
+      transportError = transportError.augmentDescription(trailers.toString());
+    } else {
+      transportError = checkContentType(trailers);
+      transportErrorMetadata = trailers;
+    }
+    if (transportError != null) {
+      http2ProcessingFailed(transportError, transportErrorMetadata);
+    } else {
+      Status status = statusFromTrailers(trailers);
+      stripTransportDetails(trailers);
+      inboundTrailersReceived(trailers, status);
+    }
+  }
+
+  private static Status statusFromHttpStatus(Metadata metadata) {
+    Integer httpStatus = metadata.get(HTTP2_STATUS);
+    if (httpStatus != null) {
+      Status status = GrpcUtil.httpStatusToGrpcStatus(httpStatus);
+      return status.isOk() ? status
+          : status.augmentDescription("extracted status from HTTP :status " + httpStatus);
+    }
+    return null;
+  }
+
+  /**
+   * Extract the response status from trailers.
+   */
+  private static Status statusFromTrailers(Metadata trailers) {
+    Status status = trailers.get(Status.CODE_KEY);
+    if (status == null) {
+      status = statusFromHttpStatus(trailers);
+      if (status == null || status.isOk()) {
+        status = Status.UNKNOWN.withDescription("missing GRPC status in response");
+      } else {
+        status = status.withDescription(
+            "missing GRPC status, inferred error from HTTP status code");
+      }
+    }
+    String message = trailers.get(Status.MESSAGE_KEY);
+    if (message != null) {
+      status = status.augmentDescription(message);
+    }
+    return status;
+  }
+
+  /**
+   * Inspect the content type field from received headers or trailers and return an error Status if
+   * content type is invalid or not present. Returns null if no error was found.
+   */
+  @Nullable
+  private Status checkContentType(Metadata headers) {
+    if (contentTypeChecked) {
+      return null;
+    }
+    contentTypeChecked = true;
+    String contentType = headers.get(GrpcUtil.CONTENT_TYPE_KEY);
+    if (!GrpcUtil.isGrpcContentType(contentType)) {
+      return Status.INTERNAL.withDescription("Invalid content-type: " + contentType);
+    }
+    return null;
+  }
+
+  /**
+   * Inspect the raw metadata and figure out what charset is being used.
+   */
+  private static Charset extractCharset(Metadata headers) {
+    String contentType = headers.get(GrpcUtil.CONTENT_TYPE_KEY);
+    if (contentType != null) {
+      String[] split = contentType.split("charset=");
+      try {
+        return Charset.forName(split[split.length - 1].trim());
+      } catch (Exception t) {
+        // Ignore and assume UTF-8
+      }
+    }
+    return Charsets.UTF_8;
+  }
+
+  /**
+   * Strip HTTP transport implementation details so they don't leak via metadata into
+   * the application layer.
+   */
+  private static void stripTransportDetails(Metadata metadata) {
+    metadata.removeAll(HTTP2_STATUS);
```

Comment:
I'll need to rebase to pull in those changes, so I'll wait toward the end to do so.